게시물 목록 API 작성

simseulnyang opened the pull request:
## PR 체크리스트
아래 항목을 확인해 주세요:

- [x] 커밋 메시지가 우리의 가이드라인을 따르고 있는지 확인하세요
- [x] 변경 사항에 대한 테스트가 추가되었는지 확인하세요 (버그 수정 / 기능 추가)
- [ ] 문서가 추가되거나 업데이트되었는지 확인하세요 (버그 수정 / 기능 추가)

## PR 유형
이 PR은 어떤 종류의 변경을 가져오나요?

<!-- 이 PR에 해당하는 것을 "x"로 표시하세요. -->

- [ ] 버그 수정
- [x] 새로운 기능 추가
- [ ] 코드 스타일 업데이트 (서식, 로컬 변수)
- [x] 리팩터링 (기능 변경 없음, API 변경 없음)
- [ ] 빌드 관련 변경
- [ ] CI 관련 변경
- [ ] 문서 내용 변경
- [ ] 애플리케이션 / 인프라 변경
- [ ] 기타... 설명:

## 현재 동작은 무엇인가요?
<!-- 수정 중인 현재 동작을 설명하거나 관련된 이슈에 링크하세요. -->

이슈 번호: #9 

## 새로운 동작은 무엇인가요?

## 이 PR은 호환성 변경을 도입하나요?

- [ ] 예
- [x] 아니요

<!-- 이 PR에 호환성 변경이 포함되어 있다면, 기존 응용 프로그램에 대한 영향과 마이그레이션 경로를 아래에 설명해 주세요. -->

## 기타 정보

simseulnyang committed:
Init: serializers.py 생성 #9

simseulnyang committed:
:tada:Init: urls.py 생성

simseulnyang committed:
Merge remote-tracking branch 'origin/develop' into feature/issue-009

simseulnyang committed:
:sparkles:Feat: filter library 추가 #9

simseulnyang committed:
:hammer:chore: decorator 파일 오타 수정

simseulnyang committed:
:sparkles:Feat : 게시물 목록 API 구현 #9

simseulnyang committed:
- 유저 계정의 해시태그를 통해 초기 필터링한 목록 리스트 구현
 - 게시글 타입별로 필터링할 수 있도록 기능 구현
 - created_at, updated_at, like_count, share_count, view_count 필드를 ordering_field로 설정하고 오름차순, 내림차순이 가능하도록 구현
 - title, content 내에서 단어 검색 시 검색할 수 있도록 구현
 - LimitOffsetPagination 기능을 통해 페이지 별 10개로 제한

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
:recycle:Feat: 외부 필터파일 생성 및 order_by 코드 리팩토링 #9

simseulnyang committed:
:sparkles:Feat: test code 생성 #9
- url을 restful하게 적용할 수 있도록 수정
- postlist 변수명 -> posts로 수정
- q 객체 사용하여 변수 지정
- 불필요한 if문 제거
- 반복적인 if-elif-else문 가독성있도록 수정
- ordering 부분 코드 수정
- test코드 추가 작성

simseulnyang committed:
:sparkles:Feat: test code 생성 #9

simseulnyang commented:
게시글 기능 관련한 코드를 작성했습니다.
그런데 요구사항 분석할 때만 하더라도 구현은 어느 정도 할 수 있을 거라고 생각했는데 
최대한 재혁님께서 구현해주신 코드와 비슷하게 해보려고 시도 하다 보니
너무 어렵고, 어지럽게 코드가 작성이 되어진 것 같습니다😢
리뷰 하시는 분들께 진심으로 죄송합니다 

그래도 일단 할 수 있는 부분까지 작성 했습니다. 테스트 코드도 일단 작성할 수 있는 부분까지 작성해 보았습니다.
코드 리뷰 해주시면 확인 하는 즉시 바로 수정하겠습니다.

saJaeHyukc commented on posts/views.py:
```diff
@@ -104,3 +107,104 @@ def get_statistics(self, queryset: Post, aggregation_field: str, value: str, agg
+            hashtag = request.user.username if o["hashtag"] is None else o["hashtag"]
+
+            # 유저 계정의 해시태그로 초기 필터링한 게시물 목록 가져오기
+            postlist = Post.objects.filter(user__username=hashtag)
```
postlist라는 변수명 네임을 지을 때 PEP8을 참고해보시면 단어 사이를 _ 표시하는게 좋을 뜻합니다! 이건 개인의 취향이지만 post_list도 좋지만 posts나 queryset은 어떻게 생각하시나요!  :) 

<img width="832" alt="image" src="https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/assets/104830931/e684a9b5-f32e-4d5b-80ee-172652511eac">

https://peps.python.org/pep-0008/

saJaeHyukc replied:
또한 q 객체를 사용하여 변수지정을 한다면 깔끔하게 코드 작성이 될 수 있을 것 같습니다 :) 
```
q = Q()
q = q & Q(user__username=hashtag)
q = q | Q(title__icontains=search_query) | Q(content__icontains=search_query)

queryset = Post.objects.filter(q)
```

saJaeHyukc commented on posts/views.py:
```diff
@@ -104,3 +107,104 @@ def get_statistics(self, queryset: Post, aggregation_field: str, value: str, agg
+            postlist = Post.objects.filter(user__username=hashtag)
+
+            # 검색어로 필터링
+            if search_query:
```
해당 if 문이 없어도 iccontains는 None일 경우에도 에러 없이 잘 나오게 되는데 제거하셔도 괜찮을 것 같아요!

saJaeHyukc commented on posts/views.py:
```diff
@@ -104,3 +107,104 @@ def get_statistics(self, queryset: Post, aggregation_field: str, value: str, agg
+            field_name = F(field_name).desc()
+
+        return postlist.order_by(field_name)
+
+    def get_post_type_list(self, postlist: Post, post_type: str):
+        if post_type == "facebook":
+            post_type_list = postlist.filter(post_type="facebook")
+        elif post_type == "twitter":
+            post_type_list = postlist.filter(post_type="twitter")
+        elif post_type == "instagram":
+            post_type_list = postlist.filter(post_type="instagram")
+        elif post_type == "threads":
+            post_type_list = postlist.filter(post_type="threads")
+        else:
+            raise InvalidParameterException("post_type 값을 잘못 선택하셨습니다.")
```
해당 코드 명시적으로  표현해서 좋지만! 아래 코드로 작성해보시면 더 간결하고 가독성이 좋을 것 같아요!

```
if post_type in ["facebook", "twitter", "instargram", "threads"]:
    postlist.filter(post_type=post_tyle)
else:
    raise InvalidParameterException("post_type 값을 잘못 선택하셨습니다.")
```

saJaeHyukc commented on posts/views.py:
```diff
@@ -104,3 +107,104 @@ def get_statistics(self, queryset: Post, aggregation_field: str, value: str, agg
+            raise UnknownServerErrorException(e)
+        return Response(serializer.data, status=status.HTTP_200_OK)
+
+    def apply_ordering(self, postlist, ordering):
+        """
+        정렬 기준을 적용하여 postlist 정렬합니다.
+        ordering에 "asc" 또는 "desc"를 추가하여 오름차순 또는 내림차순 정렬 가능합니다.
+
+        Args:
+            postlist: 정렬할 postlist
+            ordering: 정렬 기준과 방향 (예: created_at:desc)
+
+        Returns:
+            정렬된 postlist
+        """
+        ordering_parts = ordering.split(":")
+        field_name = ordering_parts[0]
+        direction = ordering_parts[1] if len(ordering_parts) > 1 else "desc"
+
+        if direction == "asc":
+            field_name = F(field_name).asc()
+        else:
+            field_name = F(field_name).desc()
+
+        return postlist.order_by(field_name)
```
asc나 desc로 구분하는 것도 좋은 접근인 것 같습니다! 제 생각은 처음 쿼리를 받아올 때 -created_at, created_at 와 같이 받아오면 해당 함수가 필요없지 않을까 생각이 듭니다 :) 어떻게 생각하시나요!

simseulnyang replied:
정렬 기준의 default값으로 created_at가 지정되었고
추가적인 기준으로 updated_at, view_count, like_count, share_count도 포함되어 있어서 
함수로 작성해보았어요! 

이 부분 좀 더 살펴보고 더 좋은 접근 방법이 뭐가 있는지 고민해보겠습니다😀

saJaeHyukc commented on posts/views.py:
```diff
@@ -104,3 +107,104 @@ def get_statistics(self, queryset: Post, aggregation_field: str, value: str, agg
+        },
+    )
+    @mandatories("type")
+    @optionals({"search": "search_by"}, {"ordering": "orderby"}, {"hashtag": None})
```
아래 작성했던 것을 확인해보시면 ordering을 orderby로 하시는 것 보다 -created_at, created_at으로 받아오면 좋지 않을까 싶습니다!

simseulnyang commented on posts/urls.py:
```diff
@@ -1,7 +1,8 @@
 
 urlpatterns = [
     path("statistics/", StatisticsListView.as_view(), name="statistics"),
+    path("list/", PostListView.as_view(), name="list"),
```
넵 확인했습니다! 
상세 페이지 작업 할 때도 참고하겠습니다😊

Chestnut90 commented on posts/views.py:
```diff
@@ -104,3 +108,104 @@ def get_statistics(self, queryset: Post, aggregation_field: str, value: str, agg
+            # 쿼리 매개변수 받기
+            post_type = m["type"]
+            search_keyword = request.query_params.get("search", "")
+            hashtag = request.user.username if o["hashtag"] is None else o["hashtag"]
```
요구사항에 hashtag가 없는 경우 username으로 쿼리하고 있는 경우 정확히 일치하는 hashtag를 쿼리해야 하니
filter가 둘로 나뉘어야 되지 않을까 생각됩니다!

simseulnyang replied:
재혁님께서 리뷰 주신 부분에 변수 활용한 방법이 있었는데

q = Q()
q = q & Q(user__username=hashtag)
q = q | Q(title__icontains=search_query) | Q(content__icontains=search_query)

queryset = Post.objects.filter(q)

이렇게 하면 필터를 나누지 않더라도 사용할 수 있지 않을까요?
일단 리뷰 주신 코드를 활용해보려고 코드 수정 중에 있습니다!
혹시 더 괜찮은 방법이 있는지 더 알아보도록 하겠습니다🙂

Chestnut90 reviewed (commented):
수고하셨습니다!
질문이 있어요.
1. django_filters가 어떻게 사용되는지 궁금합니다.

simseulnyang commented:
> 수고하셨습니다! 질문이 있어요.
> 
> 1. django_filters가 어떻게 사용되는지 궁금합니다.

제가 정확히 질문을 이해했는지 잘 모르겠지만 일단 아래에 답변 드립니다!

django_filters 사용 방법은
1. 먼저 설치를 해줍니다. (pip의 경우 pip install django-filter 이고 저는 poetry add django-filter 이렇게 설치했습니다.)
2. settings.py에서 설치한 패키지를 추가해줍니다. ("django_filters")
> 저는 THIRD_PARTY_APPS에 넣어두었습니다.

이 다음부터는 활용도가 많아서 방법은 개인적으로 선택해야 하는 것 같더라구요. 
[DRF 공식문서 'Filtering'] https://www.django-rest-framework.org/api-guide/filtering/
공식문서 참조해서 다양하게 사용하면 될 것 같습니다!

simseulnyang committed:
:recycle:Refact :Test code 리팩토링

simseulnyang committed:
:bug:Fix :migration test 오류 수정 #9

simseulnyang commented:
![image](https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/assets/123173170/ef8fbe61-433d-4cb7-bec3-b6bc2632fc74)

일단 swagger를 통해 다음 이미지와 같은 값을 받았습니다.

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
:recycle:Refact : PR 관련 코드 수정 #9
- optional에서 type부분 리스트로 수정
- 관련 없는 코드 삭제
- search_keyword값이 있을 경우, title, content 검색 가능하도록 코드 작성
- 함수 리턴값 설정

simseulnyang committed:
Merge branch 'develop' into feature/issue-009

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

JaeHyuckSa committed:
:bug:Fix: pre-commit 설정 변경 및 import 컨벤션 적용

- pre commit ymal파일이 CI에 적용된 설정 중 isort가 조금 달라서 변경하여 적용했습니다

simseulnyang committed:
:recycle:Refact : PR 리팩토링

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

JaeHyuckSa committed:
:recycle:Refact: 사용하지 않는 변수 제거 #9

- CI 과정 중 posts/views.py에 173번째 줄에 ordering이라는 변수를 사용하지 않아 제거 했습니다
Related to #9

simseulnyang committed:
:recycle:Refact : ordering 변수를 통한 정렬 코드 리팩토링

simseulnyang committed:
:recycle:Refact : ordering 변수를 사용하여 정렬 코드 리팩토링

simseulnyang committed:
:fire:Remove : 사용하지 않는 test코드 삭제

JaeHyuckSa committed:
:bug:Fix: 테스트 폴더 및 파일 중복 에러 해결 #9

- tests라는 파일과 폴더가 이중으로 존재하여 사용하지 않는 파일을 제거했습니다
Related to #9

JaeHyuckSa committed:
:bug:Fix: 통계 테스트 코드에 username 추가 #9

- username 필드가 추가가 안돼서 username을 추가해줬습니다.
Related to #9

JaeHyuckSa commented:
작업하시는 와중에 갑작스러운 커밋을 남기게 되어 죄송합니다. 아래는 CI 에러를 디버깅 과정입니다.
기본 설정을 제대로 하지 않아 일어난 일이라 꼼꼼하게 확인 후 작업하도록 하겠습니다. 불편을 끼쳐드려 죄송합니다. 

### 1 커밋 번호:[9afae8d](https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/pull/20/commits/9afae8dc689298549956a94d6fc0fb9393f49b33)
```
ERROR: /app/posts/views.py Imports are incorrectly sorted and/or formatted.
```

문제점: isort job에서 import 정렬이 안됐다고 에러 발생
원인: pre-commit-config.yaml에 작성되어있는 설정과 github actions에 작성되어있는 isort 설정이 다른 것이 원인
해결: github actions에 작성되어있는 isort 설정을 pre-commit에 적용하여 커밋 올려 수정

### 2 커밋번호: [ec5493e](https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/pull/20/commits/ec5493e14af3356f88419485d10507b4593f0631)
```
ImportError: 'tests' module incorrectly imported from '/Users/sajh/Documents/project/pre-onboarding/1_repo/posts/tests'. Expected '/Users/sajh/Documents/project/pre-onboarding/1_repo/posts'. Is this module globally installed?
```

문제점: 테스트를 실행시키려고 했으나 posts 앱에서 테스트 파일들을 못찾음
원인: tests 파일과 tests 폴더가 이중으로 존재하여 인식 못한 것이 원인
해결: 사용하지 않은 tests 파일을 제거

### 3 커밋번호: [70c494f](https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/pull/20/commits/70c494fb64dfc2712d095314613ce82e8051a221)
```
django.db.utils.IntegrityError: duplicate key value violates unique constraint "users_username_key"
DETAIL:  Key (username)=() already exists.
```

문제점: 테스트를 실행 도중 username필드값이 unique값이지만 중복으로 처리됨
원인: 기존에 작성되어있던 테스트 코드에서 유저 모델 중 username이 추가됨에 따라 작성하지 않아 null값으로 동일하게 저장되어 중복처리가 된것이 원인
해결: 기존 테스트 코드에 username을 각각 다른 값으로 저장

JaeHyuckSa commented:
<img width="439" alt="image" src="https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service/assets/104830931/0b24dc02-1424-45f3-860f-0842778ef480">

추가적으로 PostListSerializer로 query를 받아오고 있습니다. response값과 동일하게 받아오는 것으로 확인되는데 구분이 잘 안돼서PostListQuerySerialzier를 따로 만들어 적용하는 것은 어떨까요?

simseulnyang commented:
> <img alt="image" width="439" src="https://user-images.githubusercontent.com/104830931/278705822-0b24dc02-1424-45f3-860f-0842778ef480.png">
> 추가적으로 PostListSerializer로 query를 받아오고 있습니다. response값과 동일하게 받아오는 것으로 확인되는데 구분이 잘 안돼서PostListQuerySerialzier를 따로 만들어 적용하는 것은 어떨까요?

이 부분은 제가 serializer 활용하는 방법이 부족해서 하나로 사용하게 되었습니다.
재혁님께서 말씀해주신 부분에 대해 한번 살펴보고 적용해볼게요!

simseulnyang committed:
:recycle:Refact : 게시물 목록 조회 PR 코드 리팩토링

- PostFilter의 fields에 내림차순 관련 항목 추가
- PostListSerializer에 content 데이터 출력 글자 수 제한 추가
- PostQuerySerializer 구현
- query_serializer 변경
- ordering관련 filter 객체 생성 및 적용
- hashtag가 유저 계정과 동일한지 여부 판별 -> 조건에 따른 필터링 구현
- post_type관련하여 선택 전 모든 타입별 게시글이 조회되도록 설정

simseulnyang committed:
:recycle:Refact : 게시물 목록 PR 코드 리팩토링

simseulnyang committed:
:hammer:chore : git commit & push오류 수정

simseulnyang committed:
Merge branch 'develop' into feature/issue-009

simseulnyang committed:
:bug:Fix : 중복된 username 삭제

simseulnyang committed:
Merge branch 'feature/issue-009' of https://github.com/Django-Wanted-Internship-3-Team/repo1_feed-service into feature/issue-009

simseulnyang committed:
:recicle:Refactor : CI 에러 관련 PR 코드 수정

simseulnyang committed:
:recycle:Refact : PR 코드 리팩토링

- post_type에 따라 필터링 된 게시물 관련 코드 리팩토링 및 관련함수 삭제
- ordering 관련 잘못된 값 선택 시 에러메세지 출력할 수 있도록 리팩토링
- 테스트코드 수정

simseulnyang committed:
:bug:Fix : 중복된 import 삭제

- user model에 대해 import를 2번 작성하여 중복되어 있음을 확인. 불필요한
  부분을 삭제했습니다.

simseulnyang commented:
반복적인 에러 발생으로 PR이 많이 늦어진 점 죄송합니다.. 일단 PR 후 진행되는 테스트는 다 통과했는데
코드리뷰 받고 코드 수정한 부분도 있지만 에러 확인하면서 수정한 부분도 있습니다.
그래서 코드 리뷰가 필요할 것 같습니다. 
확인 부탁드립니다!

코드 수정 파일
- posts/filters.py
- posts/serializers.py
- posts/views.py
- posts/tests/views/test_posts_list_view.py

요구 사항에 있어 추가한 내용
- content 최대 출력 글자 수 제한 (-> serializer의 PostListSerializer 부분에 추가했습니다.)

Chestnut90 reviewed (approved):
고생하셨습니다!

Django에서 비즈니스 로직 작성에 대한 포스트가 있어서 공유합니다.
https://americanopeople.tistory.com/304
